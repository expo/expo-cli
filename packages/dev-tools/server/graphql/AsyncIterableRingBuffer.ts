import { $$asyncIterator } from 'iterall';

export default class AsyncIterableRingBuffer {
  buffer: any[] = [];
  _startItem: number = 0;
  _endItem: number = 0;
  _pushResolves: Function[] = [];

  constructor(public size: number) {}

  all(): number[] {
    return [...this.buffer];
  }

  allWithCursor(): { item: any; cursor: number }[] {
    return this.buffer.map((item, i) => ({
      item,
      cursor: this._startItem + i,
    }));
  }

  getNextCursor(cursor?: number): number {
    if (cursor != null && cursor >= this._startItem) {
      return cursor + 1;
    }
    return this._startItem;
  }

  getLastCursor(): number {
    return this._endItem;
  }

  async get(cursor: number) {
    if (this._endItem > cursor) {
      const adjustedCursor = cursor - this._startItem;
      return this.buffer[adjustedCursor];
    } else {
      return new Promise(resolve => {
        this._pushResolves.push(resolve);
      });
    }
  }

  length(): number {
    return this.buffer.length;
  }

  push(item: any): void {
    this.buffer.push(item);
    this._endItem++;
    this._pushResolves.forEach(resolve => resolve(item));
    this._pushResolves = [];

    if (this.buffer.length > this.size) {
      this.buffer.shift();
      this._startItem++;
    }
  }

  filterWithCursor(
    filter: (item: any, cursor: number) => number
  ): { cursor?: number; items: any[] } {
    let cursor: number | undefined;
    const items = this.buffer.filter((item, i) => {
      cursor = this._startItem + i;
      return filter(item, cursor);
    });
    return {
      cursor,
      items,
    };
  }

  getIterator(cursor?: number): Record<string, any> {
    let buffer = this;
    let iterableCursor = cursor;
    return {
      async next() {
        iterableCursor = buffer.getNextCursor(iterableCursor);
        const value = await buffer.get(iterableCursor);
        return {
          value: {
            ...value,
            cursor: iterableCursor,
          },
          done: false,
        };
      },

      [$$asyncIterator]() {
        return this;
      },
    };
  }
}
