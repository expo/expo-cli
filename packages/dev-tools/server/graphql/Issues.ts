import { EventEmitter } from 'events';

export type Issue = Record<string, any>;

export default class Issues extends EventEmitter {
  issues: Record<string, Issue> = {};

  addIssue(issueId: string, issue: Issue): void {
    let newIssue = false;
    if (!this.issues[issueId]) {
      newIssue = true;
    }
    this.issues[issueId] = {
      ...issue,
      id: issueId,
    };
    if (newIssue) {
      this.emit('ADDED', this.issues[issueId]);
    } else {
      this.emit('UPDATED', this.issues[issueId]);
    }
  }

  clearIssue(issueId: string): void {
    const issue = this.issues[issueId];
    if (issue) {
      delete this.issues[issueId];
      this.emit('DELETED', issue);
    }
  }

  getIssueList(): { cursor: string; node: Issue }[] {
    return Object.keys(this.issues).map(key => ({
      cursor: key,
      node: this.issues[key],
    }));
  }
}
