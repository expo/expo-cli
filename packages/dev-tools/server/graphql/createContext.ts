import uniqBy from 'lodash/uniqBy';
import { $$asyncIterator } from 'iterall';
import eventEmitterToAsyncIterator from '../asynciterators/eventEmitterToAsyncIterator';

export const ISSUES_SOURCE = {
  __typename: 'Issues',
  id: 'Source:issues',
  name: 'Issues',
};
export const PROCESS_SOURCE = {
  __typename: 'Process',
  id: 'Source:metro',
  name: 'Metro Bundler',
};
const DEFAULT_SOURCES = [ISSUES_SOURCE, PROCESS_SOURCE];

export default function createContext({ projectDir, messageBuffer, layout, issues }: any): any {
  return {
    getCurrentProject(): { projectDir: string } {
      return {
        projectDir,
      };
    },
    getMessageIterator(cursor: number) {
      return messageBuffer.getIterator(cursor);
    },
    getMessageEdges(source?: any) {
      if (source) {
        if (source.id === ISSUES_SOURCE.id) {
          return issues.getIssueList();
        }
        return flattenMessagesFromBuffer(messageBuffer, source.id);
      }
      return flattenMessagesFromBuffer(messageBuffer);
    },
    getMessageConnection(source: any) {
      const edges = this.getMessageEdges(source);

      let unreadCount = 0;
      let lastReadCursor = null;
      if (source) {
        ({ unreadCount, lastReadCursor } = extractReadInfo(layout.get(), source.id, edges));
      }

      return {
        count: edges.length,
        unreadCount,
        edges,
        // on-demand mapping
        nodes: () => edges.map(({ node }: any) => node),
        pageInfo: {
          hasNextPage: false,
          lastReadCursor,
          lastCursor: edges.length > 0 ? edges[edges.length - 1].cursor : null,
        },
      };
    },
    getIssuesSource() {
      return ISSUES_SOURCE;
    },
    getProcessSource() {
      return PROCESS_SOURCE;
    },
    getSourceById(id: string): any {
      const allSources = this.getSources();
      return allSources.find((source: any) => source.id === id);
    },
    getSources(): any[] {
      const chunks = messageBuffer.all().filter(({ node }: any) => node && node.tag === 'device');
      const devices = uniqBy(chunks, ({ node }) => node.deviceId).map(({ node }: any) => ({
        __typename: 'Device',
        id: node.deviceId,
        name: node.deviceName,
      }));
      return DEFAULT_SOURCES.concat(devices);
    },
    getProjectManagerLayout() {
      return layout.get();
    },
    setProjectManagerLayout(newLayout: any) {
      newLayout.sources.forEach((sourceId: string) => {
        this.setLastRead(sourceId);
      });
      layout.set(newLayout);
    },
    setLastRead(sourceId: string, lastReadCursor: any) {
      if (!lastReadCursor) {
        const source = this.getSourceById(sourceId);
        const edges = this.getMessageEdges(source);
        if (edges.length === 0) {
          return;
        } else {
          lastReadCursor = edges[edges.length - 1].cursor;
        }
      }
      layout.setLastRead(sourceId, lastReadCursor.toString());
    },
    clearMessages(sourceId: string): void {
      messageBuffer.push({
        type: 'CLEARED',
        sourceId,
      });
    },
    getIssueIterator() {
      // @ts-ignore
      const iterator = eventEmitterToAsyncIterator(issues, ['ADDED', 'UPDATED', 'DELETED']);

      return {
        async next() {
          const { value, done } = await iterator.next();
          return {
            value: {
              type: value.eventName,
              node: value.event,
            },
            done,
          };
        },
        return(): any {
          // @ts-ignore
          return iterator.return();
        },
        throw(error: Error): any {
          // @ts-ignore
          return iterator.throw(error);
        },
        [$$asyncIterator]() {
          return this;
        },
      };
    },
  };
}

function flattenMessagesFromBuffer(buffer: any, sourceId?: string) {
  const items = buffer.allWithCursor();
  const itemsById = new Map();
  const flattenedItems: any[] = [];
  for (let i = items.length - 1; i >= 0; i--) {
    const { cursor, item } = items[i];
    if (sourceId && item.sourceId !== sourceId) {
      continue;
    }
    if (item.type === 'CLEARED') {
      break;
    }
    const { node } = item;
    if (!itemsById.has(node.id)) {
      const element = { cursor, node };
      itemsById.set(node.id, element);
      flattenedItems.unshift(element);
    } else {
      itemsById.get(node.id).cursor = cursor;
    }
  }
  return flattenedItems;
}

function extractReadInfo(layout: any, sourceId: any, items: any[]) {
  let lastReadCursor = layout.sourceLastReads[sourceId];
  let unreadCount;
  if (!lastReadCursor) {
    lastReadCursor = items[0] && items[0].cursor;
    unreadCount = items.length;
  } else {
    const index = items.findIndex(({ cursor }) => cursor.toString() === lastReadCursor);
    unreadCount = items.length - index - 1;
  }
  return {
    lastReadCursor,
    unreadCount,
  };
}
