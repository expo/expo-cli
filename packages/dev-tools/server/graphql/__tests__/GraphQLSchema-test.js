import { graphql, subscribe, parse } from 'graphql';
import schema from '../GraphQLSchema';
import createContext from '../createContext';
import AsyncIterableRingBuffer from '../AsyncIterableRingBuffer';
import Issues from '../Issues';

jest.mock('@expo/xdl');

const MOCK_LOGS = [
  {
    type: 'ADDED',
    sourceId: 'Source:metro',
    node: {
      name: 'expo',
      hostname: 'freiksenet-laptop',
      pid: 29613,
      type: 'project',
      project: '/home/freiksenet/Work/expo/test-project-1',
      level: 30,
      tag: 'metro',
      msg: 'Starting Metro Bundler on port 19003.',
      time: '2018-05-07T13:01:14.738Z',
      v: 0,
      _id: 3,
      id: 3,
    },
  },
  {
    type: 'ADDED',
    sourceId: 'ebe8cd5c-9aab-48e9-9d00-0ee0e9ec3710',
    node: {
      name: 'expo',
      hostname: 'example-laptop',
      pid: 31044,
      type: 'project',
      project: '/home/test/expo/test-project-1',
      level: 30,
      _id: '74accc01-a61d-461f-b55e-f98fb629422c:0',
      tag: 'device',
      deviceId: 'ebe8cd5c-9aab-48e9-9d00-0ee0e9ec3710',
      deviceName: 'Pixel',
      groupDepth: 0,
      shouldHide: undefined,
      includesStack: false,
      msg: 'Running application "main" with appParams:',
      time: '2018-05-07T13:07:59.594Z',
      v: 0,
      id: '74accc01-a61d-461f-b55e-f98fb629422c:0',
    },
  },
  {
    type: 'ADDED',
    sourceId: 'Source:metro',
    node: {
      name: 'expo',
      hostname: 'freiksenet-laptop',
      pid: 29613,
      type: 'project',
      project: '/home/freiksenet/Work/expo/test-project-1',
      level: 30,
      tag: 'metro',
      msg: 'Metro Bundler ready.',
      time: '2018-05-07T13:01:14.878Z',
      v: 0,
      _id: 4,
      id: 4,
      _metroEventType: 'METRO_INITIALIZE_STARTED',
    },
  },
  {
    type: 'ADDED',
    sourceId: 'Source:metro',
    node: {
      name: 'expo',
      hostname: 'freiksenet-laptop',
      pid: 29613,
      type: 'project',
      project: '/home/freiksenet/Work/expo/test-project-1',
      level: 30,
      tag: 'expo',
      msg: 'Tunnel connected.',
      time: '2018-05-07T13:01:16.117Z',
      v: 0,
      _id: 6,
      id: 6,
    },
  },
];

const MOCK_BUNDLE_LOGS = [
  {
    type: 'ADDED',
    sourceId: 'Source:metro',
    node: {
      name: 'expo',
      hostname: 'freiksenet-laptop',
      pid: 31044,
      type: 'project',
      project: '/home/freiksenet/Work/expo/test-project-1',
      level: 30,
      tag: 'metro',
      msg: 'Building JavaScript bundle',
      time: '2018-05-07T13:07:56.540Z',
      v: 0,
      _id: 7,
      id: 7,
      _metroEventType: 'BUILD_PROGRESS',
      progress: 0,
      duration: 0,
    },
  },
  {
    type: 'UPDATED',
    sourceId: 'Source:metro',
    node: {
      name: 'expo',
      hostname: 'freiksenet-laptop',
      pid: 31044,
      type: 'project',
      project: '/home/freiksenet/Work/expo/test-project-1',
      level: 30,
      tag: 'metro',
      msg: 'Building JavaScript bundle',
      time: '2018-05-07T13:07:56.540Z',
      v: 0,
      _id: 7,
      id: 7,
      _metroEventType: 'BUILD_PROGRESS',
      progress: 50,
      duration: 0,
    },
  },
];

const MOCK_ISSUE = {
  name: 'expo',
  hostname: 'freiksenet-laptop',
  pid: 17465,
  type: 'project',
  project: '/home/freiksenet/Work/expo/test-project-1',
  level: 40,
  tag: 'expo',
  issueId: 'fake-issue',
  msg: 'Warning: Fake error.',
  time: '2018-05-07T13:07:56.540Z',
  v: 0,
  id: 'fake-issue',
};

const fullQuery = `
query IndexPageQuery {
  currentProject {
    projectDir
    manifestUrl
    settings {
      hostType
    }
    config {
      name
      description
      slug
    }
    sources {
      __typename
      id
      name
      messages {
        count
        nodes {
          id
          __typename
          msg
          time
        }
      }
    }
    messages {
      pageInfo {
        lastCursor
      }
    }
  }
  userSettings {
    sendTo
  }
  projectManagerLayout {
    __typename
    selected {
      id
    }
    sources {
      id
    }
  }
}`;

const messageQuery = `
query MessageQuery {
  currentProject {
    sources {
      messages {
        count
        nodes {
          id
          __typename
          msg
          time
          ... on BuildProgress {
            progress
          }
        }
      }
    }
  }
}
`;

const subscriptionQuery = parse(`
subscription MessageSubscription($after: String!) {
  messages(after: $after) {
    type
    node {
      id
      __typename
      msg
      time
      source {
        id
      }
      ... on BuildProgress {
        progress
      }
    }
  }
}
`);

let logBuffer;
let context;
let issues;
beforeEach(() => {
  logBuffer = new AsyncIterableRingBuffer();
  issues = new Issues();
  context = () =>
    createContext({
      projectDir: 'test-project-dir',
      messageBuffer: logBuffer,
      layout: {
        get() {
          return {
            selected: null,
            sources: null,
            sourceLastReads: {},
          };
        },
      },
      issues,
    });
});

test('full query', async () => {
  for (const log of MOCK_LOGS) {
    logBuffer.push(log);
  }
  issues.addIssue(MOCK_ISSUE.id, MOCK_ISSUE);
  let result = await graphql({ schema, source: fullQuery, contextValue: context() });
  expect(result).toMatchSnapshot();
});

test('message queries', async () => {
  for (const log of MOCK_LOGS.slice(0, 1)) {
    logBuffer.push(log);
  }

  let result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  for (const log of MOCK_LOGS.slice(1)) {
    logBuffer.push(log);
  }

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  logBuffer.push(MOCK_BUNDLE_LOGS[0]);

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  logBuffer.push(MOCK_BUNDLE_LOGS[1]);

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  issues.addIssue(MOCK_ISSUE.id, MOCK_ISSUE);

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  issues.addIssue(MOCK_ISSUE.id, MOCK_ISSUE);

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();

  issues.clearIssue(MOCK_ISSUE.id);

  result = await graphql({ schema, source: messageQuery, contextValue: context() });
  expect(result).toMatchSnapshot();
});

test('subscriptions', async () => {
  for (const log of MOCK_LOGS.slice(0, 1)) {
    logBuffer.push(log);
  }

  const queryResult = await graphql({ schema, source: fullQuery, contextValue: context() });
  expect(queryResult.errors).toBeUndefined();
  const cursor = queryResult.data.currentProject.messages.pageInfo.cursor;

  for (const log of MOCK_LOGS.slice(1)) {
    logBuffer.push(log);
  }

  const subscription = await subscribe({
    schema,
    document: subscriptionQuery,
    contextValue: context(),
    variableValues: { after: `${cursor}` },
  });

  issues.addIssue(MOCK_ISSUE.id, MOCK_ISSUE);

  logBuffer.push(MOCK_BUNDLE_LOGS[0]);
  logBuffer.push(MOCK_BUNDLE_LOGS[1]);

  issues.addIssue(MOCK_ISSUE.id, MOCK_ISSUE);

  issues.clearIssue(MOCK_ISSUE.id);
  issues.clearIssue('some-other-fake-id');

  const result = [];
  // Note that this expects 6 items. If there are less, then it will timeout
  // if there are more, then it won't get the ones after the first 6
  while (result.length < 9) {
    result.push((await subscription.next()).value);
  }

  expect(result).toMatchSnapshot();
});
