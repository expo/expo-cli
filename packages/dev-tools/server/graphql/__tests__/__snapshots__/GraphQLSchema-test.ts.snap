// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full query 1`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "config": null,
      "manifestUrl": "exp://mock-manifest-url",
      "messages": Object {
        "pageInfo": Object {
          "lastCursor": "3",
        },
      },
      "projectDir": "test-project-dir",
      "settings": Object {
        "hostType": "lan",
      },
      "sources": Array [
        Object {
          "__typename": "Issues",
          "id": "Source:issues",
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "Issue",
                "id": "Issue:fake-issue",
                "msg": "Warning: Fake error.",
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
          "name": "Issues",
        },
        Object {
          "__typename": "Process",
          "id": "Source:metro",
          "messages": Object {
            "count": 3,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
            ],
          },
          "name": "Metro Bundler",
        },
        Object {
          "__typename": "Device",
          "id": "ebe8cd5c-9aab-48e9-9d00-0ee0e9ec3710",
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
          "name": "Pixel",
        },
      ],
    },
    "projectManagerLayout": Object {
      "__typename": "ProjectManagerLayout",
      "selected": null,
      "sources": Array [],
    },
    "userSettings": Object {
      "sendTo": "fake-send-to",
    },
  },
}
`;

exports[`message queries 1`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 0,
            "nodes": Array [],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 2`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 0,
            "nodes": Array [],
          },
        },
        Object {
          "messages": Object {
            "count": 3,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 3`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 0,
            "nodes": Array [],
          },
        },
        Object {
          "messages": Object {
            "count": 4,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
              Object {
                "__typename": "BuildProgress",
                "id": "7",
                "msg": "Building JavaScript bundle",
                "progress": 0,
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 4`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 0,
            "nodes": Array [],
          },
        },
        Object {
          "messages": Object {
            "count": 4,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
              Object {
                "__typename": "BuildProgress",
                "id": "7",
                "msg": "Building JavaScript bundle",
                "progress": 50,
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 5`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "Issue",
                "id": "Issue:fake-issue",
                "msg": "Warning: Fake error.",
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 4,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
              Object {
                "__typename": "BuildProgress",
                "id": "7",
                "msg": "Building JavaScript bundle",
                "progress": 50,
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 6`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "Issue",
                "id": "Issue:fake-issue",
                "msg": "Warning: Fake error.",
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 4,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
              Object {
                "__typename": "BuildProgress",
                "id": "7",
                "msg": "Building JavaScript bundle",
                "progress": 50,
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`message queries 7`] = `
Object {
  "data": Object {
    "currentProject": Object {
      "sources": Array [
        Object {
          "messages": Object {
            "count": 0,
            "nodes": Array [],
          },
        },
        Object {
          "messages": Object {
            "count": 4,
            "nodes": Array [
              Object {
                "__typename": "LogMessage",
                "id": "3",
                "msg": "Starting Metro Bundler on port 19003.",
                "time": "2018-05-07T13:01:14.738Z",
              },
              Object {
                "__typename": "MetroInitializeStarted",
                "id": "4",
                "msg": "Metro Bundler ready.",
                "time": "2018-05-07T13:01:14.878Z",
              },
              Object {
                "__typename": "LogMessage",
                "id": "6",
                "msg": "Tunnel connected.",
                "time": "2018-05-07T13:01:16.117Z",
              },
              Object {
                "__typename": "BuildProgress",
                "id": "7",
                "msg": "Building JavaScript bundle",
                "progress": 50,
                "time": "2018-05-07T13:07:56.540Z",
              },
            ],
          },
        },
        Object {
          "messages": Object {
            "count": 1,
            "nodes": Array [
              Object {
                "__typename": "DeviceMessage",
                "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
                "msg": "Running application \\"main\\" with appParams:",
                "time": "2018-05-07T13:07:59.594Z",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`subscriptions 1`] = `
Array [
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "Issue",
          "id": "Issue:fake-issue",
          "msg": "Warning: Fake error.",
          "source": Object {
            "id": "Source:issues",
          },
          "time": "2018-05-07T13:07:56.540Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "LogMessage",
          "id": "3",
          "msg": "Starting Metro Bundler on port 19003.",
          "source": Object {
            "id": "Source:metro",
          },
          "time": "2018-05-07T13:01:14.738Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "Issue",
          "id": "Issue:fake-issue",
          "msg": "Warning: Fake error.",
          "source": Object {
            "id": "Source:issues",
          },
          "time": "2018-05-07T13:07:56.540Z",
        },
        "type": "UPDATED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "DeviceMessage",
          "id": "74accc01-a61d-461f-b55e-f98fb629422c:0",
          "msg": "Running application \\"main\\" with appParams:",
          "source": Object {
            "id": "ebe8cd5c-9aab-48e9-9d00-0ee0e9ec3710",
          },
          "time": "2018-05-07T13:07:59.594Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "Issue",
          "id": "Issue:fake-issue",
          "msg": "Warning: Fake error.",
          "source": Object {
            "id": "Source:issues",
          },
          "time": "2018-05-07T13:07:56.540Z",
        },
        "type": "DELETED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "MetroInitializeStarted",
          "id": "4",
          "msg": "Metro Bundler ready.",
          "source": Object {
            "id": "Source:metro",
          },
          "time": "2018-05-07T13:01:14.878Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "LogMessage",
          "id": "6",
          "msg": "Tunnel connected.",
          "source": Object {
            "id": "Source:metro",
          },
          "time": "2018-05-07T13:01:16.117Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "BuildProgress",
          "id": "7",
          "msg": "Building JavaScript bundle",
          "progress": 0,
          "source": Object {
            "id": "Source:metro",
          },
          "time": "2018-05-07T13:07:56.540Z",
        },
        "type": "ADDED",
      },
    },
  },
  Object {
    "data": Object {
      "messages": Object {
        "node": Object {
          "__typename": "BuildProgress",
          "id": "7",
          "msg": "Building JavaScript bundle",
          "progress": 50,
          "source": Object {
            "id": "Source:metro",
          },
          "time": "2018-05-07T13:07:56.540Z",
        },
        "type": "UPDATED",
      },
    },
  },
]
`;
